eclair-core/src/main/scala/fr/acinq/eclair/tor/TorProtocolHandler.scala

```diff
@@ -0,0 +1,317 @@
+
+import akka.actor.{Actor, ActorLogging, ActorRef, Props, Stash, Status}
+import akka.io.Tcp.Connected
+import akka.util.ByteString
+import fr.acinq.eclair.tor.TorProtocolHandler.ProtocolVersion
+import fr.acinq.eclair.randomBytes
+import javax.crypto.Mac
+import javax.crypto.spec.SecretKeySpec
+import javax.xml.bind.DatatypeConverter
+
+import scala.concurrent.Promise
+import scala.util.Random
+
+case class TorException(msg: String) extends RuntimeException(msg)
+
+class TorProtocolHandler(protocolVersion: ProtocolVersion,
+                         privateKeyPath: String,
+                         virtualPort: Int,
+                         targetPorts: Seq[Int],
+                         onionAdded: Option[Promise[OnionAddress]],
+                         clientNonce: Option[Array[Byte]]
+                        ) extends Actor with Stash with ActorLogging {
+
+  import TorProtocolHandler._
+
+  private val ServerKey: Array[Byte] = "Tor safe cookie authentication server-to-controller hash".getBytes()
+  private val ClientKey: Array[Byte] = "Tor safe cookie authentication controller-to-server hash".getBytes()
+
+  private var receiver: ActorRef = _
+
+  private var protoInfo: ProtocolInfo = _
+  private var clientHash: Array[Byte] = _
+  private var keyStr: String = _
+  private var portStr: String = _
+
+  private var address: Option[OnionAddress] = None
+
+  private val nonce: Array[Byte] = clientNonce.getOrElse(randomBytes(32))
+
+  override def receive: Receive = {
+    case Connected(_, _) =>
+      receiver = sender()
+      context become protocolInfo
+      self ! SendNextCommand
+  }
+
+  def protocolInfo: Receive = {
+    case SendNextCommand =>
+      sendCommand("PROTOCOLINFO 1")
+    case data: ByteString => handleExceptions {
+      val res = parseResponse(readResponse(data))
+      protoInfo = ProtocolInfo(
+        methods = res.getOrElse("METHODS", throw TorException("Tor auth methods not found")),
+        cookieFile = unquote(res.getOrElse("COOKIEFILE", throw TorException("Tor cookie file not found"))),
+        version = unquote(res.getOrElse("Tor", throw TorException("Tor version not found"))))
+      log.info(s"Tor version ${protoInfo.version}")
+      if (!protocolVersion.supportedBy(protoInfo.version)) {
+        throw TorException(s"Tor version ${protoInfo.version} does not support protocol $protocolVersion")
+      }
+      context become authChallenge
+      self ! SendNextCommand
+    }
+  }
+
+  def authChallenge: Receive = {
+    case SendNextCommand =>
+      sendCommand(s"AUTHCHALLENGE SAFECOOKIE ${hex(nonce)}")
+    case data: ByteString => handleExceptions {
+      val res = parseResponse(readResponse(data))
+      clientHash = computeClientHash(
+        res.getOrElse("SERVERHASH", throw TorException("Tor server hash not found")),
+        res.getOrElse("SERVERNONCE", throw TorException("Tor server nonce not found"))
+      )
+      context become authenticate
+      self ! SendNextCommand
+    }
+  }
```

Comment:
Did you consider not using `SendNextCommand` and instead sending the next command just before the call to `context. become` ? 

For example:
```scala
  sendCommand(s"AUTHENTICATE ${hex(clientHash)}")
  context become authenticate
````

Reply:
Good idea! 👍